pkg/util/kubernetes/apiserver/hpa_controler.go

```diff
@@ -0,0 +1,312 @@
+	queue workqueue.RateLimitingInterface
+
+	// used in unit tests to wait until endpoints are synced
+	autoscalers chan interface{}
+
+	toStore   hpaToStoreGlobally
+	hpaProc   *hpa.HPAProcessor
+	store     custommetrics.Store
+	clientSet kubernetes.Interface
+	poller    Poller
+	le        LeaderElectorItf
+}
+
+// NewAutoscalerController returns a new AutoscalerController
+func NewAutoscalerController(client kubernetes.Interface, le LeaderElectorItf, dogCl hpa.DatadogClient, autoscalingInformer autoscalersinformer.HorizontalPodAutoscalerInformer) *AutoscalersController {
+	var err error
+	h := &AutoscalersController{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "autoscalers"),
+	}
+
+	h.poller = Poller{
+		gcPeriodSeconds: config.Datadog.GetInt("hpa_watcher_gc_period"),
+		refreshPeriod:   config.Datadog.GetInt("external_metrics_provider.polling_freq"),
+		batchWindow:     config.Datadog.GetInt("external_metrics_provider.batch_window"),
+	}
+
+	// Setup the client to process the HPA and metrics
+	h.hpaProc, err = hpa.NewHPAProcessor(dogCl)
+	if err != nil {
+		log.Errorf("Could not instantiate the HPA Processor: %v", err.Error())
+		return nil
+	}
+	h.clientSet = client
+	h.le = le // only trigger GC and updateExternalMetrics by the Leader.
+
+	datadogHPAConfigMap := custommetrics.GetConfigmapName()
+	h.store, err = custommetrics.NewConfigMapStore(client, common.GetResourcesNamespace(), datadogHPAConfigMap)
+	if err != nil {
+		log.Errorf("Could not instantiate the local store for the External Metrics %v", err)
+		return nil
+	}
+
+	autoscalingInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    h.addAutoscaler,
+			UpdateFunc: h.updateAutoscaler,
+			DeleteFunc: h.deleteAutoscaler,
+		},
+	)
+	h.autoscalersLister = autoscalingInformer.Lister()
+	h.autoscalersListerSynced = autoscalingInformer.Informer().HasSynced
+	return h
+}
+
+func (h *AutoscalersController) Run(stopCh <-chan struct{}) {
+	defer h.queue.ShutDown()
+
+	log.Infof("Starting HPA Controller ... ")
+	defer log.Infof("Stopping HPA Controller")
+
+	if !cache.WaitForCacheSync(stopCh, h.autoscalersListerSynced) {
+		return
+	}
+
+	h.processingLoop()
+
+	go wait.Until(h.worker, time.Second, stopCh)
+	<-stopCh
+}
+
+// processingLoop is a second go routing that schedules the garbage collection and the refreshing of external metrics
+// in the GlobalStore.
+func (c *AutoscalersController) processingLoop() {
+	tickerHPARefreshProcess := time.NewTicker(time.Duration(c.poller.refreshPeriod) * time.Second)
+	gcPeriodSeconds := time.NewTicker(time.Duration(c.poller.gcPeriodSeconds) * time.Second)
+	batchFreq := time.NewTicker(time.Duration(c.poller.batchWindow) * time.Second)
+
+	go func() {
+		for {
+			select {
+			case <-tickerHPARefreshProcess.C:
+				if !c.le.IsLeader() {
+					c.toStore.m.Lock()
+					c.toStore.data = nil
+					c.toStore.m.Unlock()
+					continue
```

Comment:
I'm confused by what is going on here.

```
c.toStore.m.Lock()
c.toStore.data = nil
c.toStore.m.Unlock()
```

 I think the `batchFreq` ticker should be in charge of clearing out the cache if we are not currently the leader.